accesscontroltool-startuphook-bundle/src/main/java/biz/netcentric/cq/tools/actool/startuphook/impl/AcToolStartupHookServiceImpl.java

```diff
@@ -64,36 +81,76 @@ public void activate(BundleContext bundleContext, Config config) {
                 RuntimeHelper.getCurrentStartLevel(bundleContext),
                 isCloudReady,
-                activationMode);
+                activationMode,
+                runAsyncForMutableConent);
 
         boolean applyOnStartup = (activationMode == Config.StartupHookActivation.ALWAYS) 
                 || (isCloudReady && activationMode == Config.StartupHookActivation.CLOUD_ONLY);
 
         if(applyOnStartup) {
-            
-            try {
-
-                List<String> relevantPathsForInstallation = getRelevantPathsForInstallation();
-                LOG.info("Running AcTool with "
-                        + (relevantPathsForInstallation.isEmpty() ? "all paths" : "paths " + relevantPathsForInstallation) + "...");
-                acInstallationService.apply(null, relevantPathsForInstallation.toArray(new String[relevantPathsForInstallation.size()]),
-                        true);
-                LOG.info("AC Tool Startup Hook done. (start level " + RuntimeHelper.getCurrentStartLevel(bundleContext) + ")");
 
-                copyAcHistoryToOrFromApps(isCloudReady);
-
-            } catch (RepositoryException e) {
-                LOG.error("Exception while triggering AC Tool on startup: " + e, e);
+            if (runAsyncForMutableConent && isCompositeNodeStore) {
+                runAcToolAsync(bundleContext, isCloudReady);
+            } else {
+                runAcTool(bundleContext, isCloudReady);
             }
+
         } else {
             LOG.debug("Skipping AcTool Startup Hook: activationMode: {} isCloudReady: {}", activationMode, isCloudReady);
         }
 
     }
 
+    @Deactivate
+    public void deactivate() {
+        if (jobHandlerRegistration != null) {
+            jobHandlerRegistration.unregister();
+        }
+    }
+
+    private void runAcTool(BundleContext bundleContext, boolean isCloudReady) {
+        try {
+
+            List<String> relevantPathsForInstallation = getRelevantPathsForInstallation();
+            LOG.info("Running AcTool with "
+                    + (relevantPathsForInstallation.isEmpty() ? "all paths" : "paths " + relevantPathsForInstallation) + "...");
+            acInstallationService.apply(null, relevantPathsForInstallation.toArray(new String[relevantPathsForInstallation.size()]),
+                    true);
+            LOG.info("AC Tool Startup Hook done. (start level " + RuntimeHelper.getCurrentStartLevel(bundleContext) + ")");
+
+            copyAcHistoryToOrFromApps(isCloudReady);
+
+        } catch (RepositoryException e) {
+            LOG.error("Exception while triggering AC Tool on startup: " + e, e);
+        }
+    }
+
+    private void runAcToolAsync(final BundleContext bundleContext, final boolean isCloudReady) {
+
+        final AcToolStartupHookServiceImpl startupHook = this;
+        JobConsumer acToolJobConsumer = new JobConsumer() {
```

Comment:
I would not go for sling jobs but rather for lightweight threads or sling scheduler

Reply:
IIRC the running job will prevent shutdown of the instance while the job is running while the scheduler/a thread might not?

Reply:
No, even Sling Jobs will not prevent shutdown, but the persistence make them start again the next time.

Reply:
fair enough... to keep it simple I'd just go for a thread then (the scheduler is not available during image build, so we would have to decouple via an optional service, not worth the effort, a simple thread should do the job)